<a href="https://colab.research.google.com/github/Rama389/AI-Projects/blob/main/Deep_Learning_Forward_Propogation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Deep Learning

## Forward Propagation

Procedure for calculation of Xavier weights

Input to hidden:

inputs = 2

hidden = 2 (neurons)

output = 1

Input1 (i1) = 2

Input2 (i2) = 4

Linear Activation function


In [ ]:
import numpy as np

# ----------------------------
# Xavier Initialization Helper
# ----------------------------
def xavier_init(fan_in, fan_out):
    limit = np.sqrt(6.0 / (fan_in + fan_out))
    return np.random.uniform(-limit, limit, size=(fan_in, fan_out))

# ----------------------------
# Network Setup: 2 -> 2 -> 1
# ----------------------------
np.random.seed(42)  # for reproducibility

# Inputs
i1, i2 = 2.0, 4.0
x = np.array([i1, i2])  # shape (2,)

# Xavier Initialization
W_input_hidden = xavier_init(2, 2)   # shape (2,2) -> W1..W4
W_hidden_output = xavier_init(2, 1)  # shape (2,1) -> W5..W6

# Report initialized weights
print("=== Xavier Initialized Weights ===")
print("W1, W2, W3, W4 (input -> hidden):")
print(W_input_hidden)
print("W5, W6 (hidden -> output):")
print(W_hidden_output.flatten())

# ----------------------------
# Forward Propagation (linear hidden)
# ----------------------------
# Hidden layer (linear activation, so no tanh/sigmoid)
z_h = np.dot(x, W_input_hidden)   # shape (2,) -> [h1, h2]
h1, h2 = z_h

# Output layer (linear)
z_out = np.dot(z_h, W_hidden_output).item()

# Report forward pass
print("\n=== Forward Propagation ===")
print(f"h1 = {h1:.4f}")
print(f"h2 = {h2:.4f}")
print(f"out (prediction) = {z_out:.4f}")

=== Xavier Initialized Weights ===
W1, W2, W3, W4 (input -> hidden):
[[-0.30731269  1.10402007]
 [ 0.56826678  0.24166295]]
W5, W6 (hidden -> output):
[-0.97292621 -0.97299443]

=== Forward Propagation ===
h1 = 1.6584
h2 = 3.1747
out (prediction) = -4.7025


The weights are;
w1 = -0.3, w2 = -1.1, w3 = 0.56, w4 = 0.24, w5 = -0.97, w6 = 0.24

Output = -4.7025

# Question - 4

## ReLU Activation Function

In [ ]:
import numpy as np

# given
W1, W2, W3, W4 = -0.30731343, 1.10402055, 0.56826769, 0.24166314
W5, W6 = -0.97292573, -0.97299407
i1, i2 = 2.0, 4.0

# pre-activations
h1 = W1*i1 + W2*i2
h2 = W3*i1 + W4*i2

# ReLU
h1_relu = max(0.0, h1)
h2_relu = max(0.0, h2)

# output
out = W5*h1_relu + W6*h2_relu

print("h1:", h1)
print("h2:", h2)
print("h1 (ReLU):", h1_relu)
print("h2 (ReLU):", h2_relu)
print("out:", out)


h1: 3.80145534
h2: 2.1031879399999998
h1 (ReLU): 3.80145534
h2 (ReLU): 2.1031879399999998
out: -5.744923105447414


Using ReLU activation function the output value is -5.7449